rewrite uiElementIsVisible

thot-experiment opened the pull request:
rewrite visibility checking to be more generic/cleaner as well as add functionality to check if the element is scrolled on screen for more intuitive paste-target selection, fixes #10488

the improvement is that now we check the scroll location of viable elements when pasting an image and paste the image into the topmost element that's visible **on screen**, making pasted image inputs more intuitive

previously if using controlnets/segmentation etc in i2i, the pasted image would always land in the i2i input image window, now if you scroll down to controlnet such that the i2i window is no longer visible, the pasted image will end up in the controlnet input rather than the i2i input, this simplifies the use of addons that accept image input in conjunction with i2i or when using multiple addons that accept image input in t2i

i've done a bit of testing with different addons and have not found any regressions vs the previous 1.1 functionality, and this fixes an issue in 1.2 where in some circumstances pasted images would end up in a hidden pane as per #10488

**Environment this was tested in**

 - OS: Windows 10
 - Browser: firefox, chrome
 - Graphics card: NVIDIA GV100

**Screenshots or videos of your changes**

This is not strictly a UI change as the UI still appears the exact same? I'm not sure exactly what I should provide a screenshot/video of here.

thot-experiment committed:
rewrite uiElementIsVisible

rewrite visibility checking to be more generic/cleaner as well as add functionality to check if the element is scrolled on screen for more intuitive paste-target selection

AUTOMATIC1111 commented:
Will paste stop working when you have a valid visible image input but it's scrolled out of vision?

thot-experiment committed:
split visibility method and sort instead

split out the visibility method for pasting and use a sort inside the paste handler to prioritize on-screen fields rather than targeting ONLY on screen fields

thot-experiment commented:
> Will paste stop working when you have a valid visible image input but it's scrolled out of vision?

Not anymore! :3 (as of [7b61acb](https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/10534/commits/7b61acbd35e9db43a5279a42afad3f3dc68462c9))
